java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,354 @@
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  public Glome(XECPublicKey peerKey, XECPrivateKey userPrivateKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    XECPublicKey userPublicKey = derivePublicKeyFromPrivate(userPrivateKey);
+    this.userKeys = new KeyPair(userPublicKey, userPrivateKey);
+    initMacKeys(peerKey, minPeerTagLength);
+  }
+
+  public KeyPair userKeys() {
+    return userKeys;
+  }
+
+  public XECPublicKey peerKey() {
+    return peerKey;
+  }
+
+  public XECPublicKey userPublicKey() {
+    return (XECPublicKey) userKeys.getPublic();
+  }
+
+  public XECPrivateKey userPrivateKey() {
+    return (XECPrivateKey) userKeys.getPrivate();
+  }
+
+  /**
+   * Generates a user's tag corresponding to the given message {@code msg} and the counter {@code
+   * cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the peer to the user.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [0..255] range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than {@code MAX_TAG_LENGTH}) or it's not equal to the prefix of a
+   * correct tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          String.format("The received tag has invalid length. Expected %d or more, got %d.",
+              minPeerTagLength, peerTag.length));
+    }
+
+    byte[] truncatedTag = Arrays.copyOf(generateTag(msg, cnt, this.peerMacKey), peerTag.length);
+    if (!Arrays.equals(peerTag, truncatedTag)) {
+      throw new WrongTagException("The received tag doesn't match the expected tag.");
+    }
+  }
+
+  /**
+   * Common part of the constructors. Initialises peer's public key, minimum length of a peer's tag,
+   * base point for Curve25519. Calculates user's and peer's MAC keys.
+   *
+   * @param peerKey peer's public key.
+   * @param minPeerTagLength minimum length of a peer's tag.
+   * @throws MinPeerTagLengthOutOfBoundsException if {@code minPeerTagLength} is out of [1..32]
+   * range.
+   */
+  private void initMacKeys(XECPublicKey peerKey, int minPeerTagLength)
+      throws MinPeerTagLengthOutOfBoundsException {
+    this.peerKey = peerKey;
```

Comment:
This method can fail in ways that leave the object partially updated. Try to perform validation and risky work up front, then on success, update the object members with the result as a final step.

Reply:
Thank you! Made an init block.